File: src/test/java/com/soapdogg/lispInterpreter/parser/NodeParserTest.java
package com.soapdogg.lispInterpreter.parser;

import com.soapdogg.lispInterpreter.asserter.TokenKindAsserter;
import com.soapdogg.lispInterpreter.datamodels.ParserResult;
import com.soapdogg.lispInterpreter.datamodels.Token;
import com.soapdogg.lispInterpreter.datamodels.TokenKind;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import java.util.LinkedList;
import java.util.List;
import java.util.Optional;

public class NodeParserTest {

    private ExpressionNodeParser expressionNodeParser;
    private ExpressionNodeFinisher expressionNodeFinisher;
    private AtomNodeParser atomNodeParser;

    private Token headToken;
    private List<Token> tokens;

    private NodeParser nodeParser;

    @BeforeEach
    void setup() {
        TokenKindAsserter tokenKindAsserter = Mockito.mock(TokenKindAsserter.class);
        expressionNodeParser = Mockito.mock(ExpressionNodeParser.class);
        expressionNodeFinisher = Mockito.mock(ExpressionNodeFinisher.class);
        atomNodeParser = Mockito.mock(AtomNodeParser.class);

        headToken = Mockito.mock(Token.class);
        tokens = new LinkedList<>();
        tokens.add(headToken);

        Mockito.when(
            tokenKindAsserter.assertTokenIsNotNull(Optional.of(headToken))
        ).thenReturn(headToken);

        Mockito.doNothing().when(tokenKindAsserter).assertTokenIsAtomOrOpen(headToken);

        nodeParser = new NodeParser(
            tokenKindAsserter,
            expressionNodeParser,
            expressionNodeFinisher,
            atomNodeParser
        );
    }

    @Test
    void parseIntoExpressionNodeTest() {
        Mockito.when(headToken.getTokenKind()).thenReturn(TokenKind.OPEN_TOKEN);

        final ParserResult result = Mockito.mock(ParserResult.class);
        Mockito.when(expressionNodeParser.parseExpressionNode(new LinkedList<>())).thenReturn(result);

        final ParserResult expected = Mockito.mock(ParserResult.class);
        Mockito.when(expressionNodeFinisher.finishParsingExpressionNode(result)).thenReturn(expected);

        final ParserResult actual = nodeParser.parseIntoNode(tokens);

        Assertions.assertEquals(expected, actual);

        Mockito.verifyNoInteractions(atomNodeParser);
    }

    @Test
    void parseIntoAtomNodeTest() {
        Mockito.when(headToken.getTokenKind()).thenReturn(TokenKind.NUMERIC_TOKEN);

        final ParserResult expected = Mockito.mock(ParserResult.class);
        Mockito.when(
            atomNodeParser.parseAtomNode(tokens)
        ).thenReturn(expected);

        final ParserResult actual = nodeParser.parseIntoNode(tokens);

        Assertions.assertEquals(expected, actual);

        Mockito.verifyNoInteractions(expressionNodeParser);
    }
}

File: src/test/java/com/soapdogg/lispInterpreter/parser/NodeParserTest.kt
package com.soapdogg.lispInterpreter.parser

import com.soapdogg.lispInterpreter.datamodels.ParserResult
import com.soapdogg.lispInterpreter.datamodels.Token
import com.soapdogg.lispInterpreter.datamodels.TokenKind
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import java.util.*

class NodeParserTest {
    private val expressionNodeParser = Mockito.mock(ExpressionNodeParser::class.java)
    private val expressionNodeFinisher = Mockito.mock(ExpressionNodeFinisher::class.java)
    private val atomNodeParser = Mockito.mock(AtomNodeParser::class.java)
    private val nodeParser = NodeParser(
        expressionNodeParser,
        expressionNodeFinisher,
        atomNodeParser
    )

    @Test
    fun parseIntoExpressionNodeTest() {
        val headToken = Mockito.mock(Token::class.java)
        val tokens = listOf(headToken)
        Mockito.`when`(headToken.tokenKind).thenReturn(TokenKind.OPEN_TOKEN)

        val result = Mockito.mock(ParserResult::class.java)
        Mockito.`when`(expressionNodeParser.parseExpressionNode(LinkedList())).thenReturn(result)

        val expected = Mockito.mock(ParserResult::class.java)
        Mockito.`when`(expressionNodeFinisher.finishParsingExpressionNode(result)).thenReturn(expected)

        val actual = nodeParser.parseIntoNode(tokens)
        Assertions.assertEquals(expected, actual)
        Mockito.verifyNoInteractions(atomNodeParser)
    }

    @Test
    fun parseIntoAtomNodeTest() {
        val headToken = Mockito.mock(Token::class.java)
        val tokens = listOf(headToken)
        Mockito.`when`(headToken.tokenKind).thenReturn(TokenKind.NUMERIC_TOKEN)

        val expected = Mockito.mock(ParserResult::class.java)
        Mockito.`when`(
            atomNodeParser.parseAtomNode(tokens)
        ).thenReturn(expected)

        val actual = nodeParser.parseIntoNode(tokens)
        Assertions.assertEquals(expected, actual)
        Mockito.verifyNoInteractions(expressionNodeParser)
    }
}